### ISSUE TITLE
[Bug] Multimodal cache hash collisions cause embedding corruption and DoS

### ISSUE THREAD

**Issue by 3em0** [sgl-project/sglang#25462]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

The multimodal embedding cache key path has several hash stability and collision issues that can cause silent embedding corruption, cache invalidation across processes, or request crashes.

Affected code paths:

- `python/sglang/srt/managers/mm_utils.py`: `data_hash`, `tensor_hash`, `hash_feature`
- `python/sglang/srt/mem_cache/multimodal_cache.py`: `combine_hashes`
- `python/sglang/srt/multimodal/processors/base_processor.py`: `precomputed_embedding` handling

Issues:

1. `data_hash` truncates SHA256 to 64 bits with `hashlib.sha256(data).digest()[:8]`, reducing the birthday-bound collision threshold to about `2^32`.
2. `tensor_hash` flattens and concatenates tensor lists without encoding per-tensor boundaries, shape, or dtype. Different tensor partitions with identical concatenated bytes produce identical hashes.
3. The `np.ndarray` branch of `hash_feature` hashes only `arr.tobytes()`, so arrays with identical bytes but different shape or dtype can collide.
4. `MultimodalCache.combine_hashes` uses Python's built-in `hash(tuple(...))`, which is not stable across independent processes because of `PYTHONHASHSEED` randomization.
5. The non-tensor list branch in `hash_feature` passes a `tuple` into `data_hash`, which calls `hashlib.sha256()` and raises `TypeError` because the input is not bytes-like.
6. `precomputed_embedding` accepts arbitrary `feature` values without sufficient validation, which makes cache-key inputs directly user controlled.

Impact:

- Deterministic cache poisoning: different multimodal inputs can reuse the same cached embedding.
- Silent incorrect outputs: a request can receive an embedding computed for a different multimodal input.
- Denial of service: non-tensor list features can crash the request path with `TypeError`.
- Cache instability: distributed workers, TP workers, or restarted processes may compute different combined cache keys for the same multimodal inputs.

### Reproduction

The following minimal script demonstrates the deterministic collision and crash cases without requiring a model server:

```python
import hashlib
import numpy as np
import torch


def data_hash(data) -> int:
    hash_bytes = hashlib.sha256(data).digest()[:8]
    return int.from_bytes(hash_bytes, byteorder="big", signed=False)


def flatten_nested_list(nested_list):
    ret = []
    for item in nested_list:
        if isinstance(item, list):
            ret.extend(flatten_nested_list(item))
        else:
            ret.append(item)
    return ret


def tensor_hash(tensor_list) -> int:
    if isinstance(tensor_list, list):
        tensor_list = flatten_nested_list(tensor_list)
        tensor_list = [x.flatten() for x in tensor_list]
        tensor = torch.concat(tensor_list)
    else:
        tensor = tensor_list
    return data_hash(tensor.numpy().tobytes())


def hash_feature(f):
    if isinstance(f, list):
        if isinstance(f[0], torch.Tensor):
            return tensor_hash(f)
        return data_hash(tuple(flatten_nested_list(f)))
    elif isinstance(f, np.ndarray):
        arr = np.ascontiguousarray(f)
        return data_hash(arr.tobytes())
    elif isinstance(f, torch.Tensor):
        return tensor_hash([f])
    return data_hash(f)


# 1. Tensor boundary loss: same bytes, different tensor partitions.
a = [torch.tensor([1.0, 2.0]), torch.tensor([3.0, 4.0, 5.0])]
b = [torch.tensor([1.0, 2.0, 3.0]), torch.tensor([4.0, 5.0])]
print("tensor_hash(a):", tensor_hash(a))
print("tensor_hash(b):", tensor_hash(b))
assert tensor_hash(a) == tensor_hash(b)


# 2. ndarray shape loss: same bytes, different shapes.
data = np.arange(16, dtype=np.float32)
x = data.reshape(16,)
y = data.reshape(4, 4)
print("hash_feature(x):", hash_feature(x))
print("hash_feature(y):", hash_feature(y))
assert hash_feature(x) == hash_feature(y)


# 3. Non-tensor list crash.
try:
    hash_feature([1, 2, 3])
except TypeError as e:
    print("TypeError:", e)
```

### Environment

```text
Python: 3.10.12
CUDA available: True
GPU 0: NVIDIA A100-SXM4-40GB
GPU 0 Compute Capability: 8.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.41
CUDA Driver Version: 580.126.20
PyTorch: 2.11.0+cu130
sglang: 0.5.12.dev476+g4df42da65
sglang-kernel: 0.4.2.post2
flashinfer_python: 0.6.11.post1
flashinfer_cubin: 0.6.11.post1
triton: 3.6.0
transformers: 5.6.0
```

**Comment by abinggo** [comment]
You have an open PR for this — #22033 from April 3rd, covering all 6 issues you raised here. It's been blocked since day one waiting for the `run-ci` label gate (the CI failures are from the missing label, not real test failures). I just left a code review there.

Pinging @yhyang201 / @mickqian directly on the PR is more likely to unblock it than reopening the issue — they've owned `mm_utils.py` and `multimodal_cache.py` recently and have context on the tensor-hashing and multimodal-cache paths.

**Comment by 3em0** [comment]
@yhyang201 @mickqian
